### PR TITLE
Track addon usage over time

### PR DESCRIPTION
**Pull request by jamesrkiger**
## Checklist

1. [ ] If you've added code that should be tested, add tests
2. [ ] If you've changed APIs, update (or create!) the documentation
3. [x] Ensure the tests pass
4. [x] Make sure that your code lints and that you've followed [our coding style](https://github.com/kobotoolbox/kpi/blob/master/CONTRIBUTING.md)
5. [x] Write a title and, if necessary, a description of your work suitable for publishing in our [release notes](https://community.kobotoolbox.org/tag/release-notes)
6. [ ] Mention any related issues in this repository (as #ISSUE) and in other repositories (as kobotoolbox/other#ISSUE)
7. [ ] Open an issue in the [docs](https://github.com/kobotoolbox/docs/issues/new) if there are UI/UX changes

## Description

The basic goal here is to enable the frontend to display how many units of addon consumption were used in the current billing period. For that, we need to track addon usage as it happens. This PR aims to do that by adding new addon usage keys to the NLPUsageCounter jsonfield dict whenever an addon's usage is incremented.  The aggregated data is served to the frontend via the service_usage/organization service_usage endpoints.

Related changes:
- Overhauled `is_organization_over_plan_limit` method to check not just whether an org is over its limit, but whether it will go over its limit with a specified usage amount.

- Updated project ownership transfer code to remove addon usage counts from transferred NLPUsageCounters, as these counts will not be relevant to a new owner.

Unrelated changes:
- Fixed tests that were already broken by earlier changes in `test_one_time_addons_api.py`

## Notes

There are still some deeper issues with how we handle usage caching that will need to be resolved before this project is completed. But I think the work I have done so far could be merged into the project branch so the PR for that cache work will be more manageable.

Credit to @noliveleger for the new `IncrementValues` ORM helper.

### CODE REVIEW

File: kobo/apps/stripe/constants.py [hunk 1]
```diff
@@ -27,10 +27,3 @@
     'storage': 'storage_bytes',
     'submission': 'submission',
 }
-
-USAGE_LIMIT_MAP_STRIPE = {
```
Comment (jamesrkiger) [inline kobo/apps/stripe/constants.py:31]: If I'm not mistaken, this alternative mapping const is no longer needed. We use asr/mt rather than 'nlp' in stripe products now, right?

File: kobo/apps/stripe/tests/test_one_time_addons_tracking.py [hunk 2]
```diff
@@ -0,0 +1,196 @@
...
+from kobo.apps.trackers.models import NLPUsageCounter
+from kobo.apps.trackers.submission_utils import (
+    create_mock_assets,
+)
+from kobo.apps.stripe.models import PlanAddOn
+from kobo.apps.trackers.utils import update_nlp_counter
+from kpi.tests.api.v2.test_api_service_usage import ServiceUsageAPIBase
+
+
+class OneTimeAddonTrackingTestCase(ServiceUsageAPIBase):
```
Comment (jamesrkiger) [inline kobo/apps/stripe/tests/test_one_time_addons_tracking.py:27]: This is still an interim test class at the moment, but I do wonder how I should clean it up once it is ready. There is a fair amount of overlap with `ServiceUsageAPIBase` and `OrganizationUsageAPITestCase`. Ideally, I think I would create a base test class with methods flexible enough to work in all three classes. But that might be better left to another PR.

File: kpi/deployment_backends/kobocat_backend.py [hunk 3]
```diff
@@ -1286,9 +1286,19 @@ def transfer_submissions_ownership(
 
     def transfer_counters_ownership(self, new_owner: 'auth.User'):
 
+        # Addon usage is not relevant to new owner, so we zero it
         NLPUsageCounter.objects.filter(
             asset=self.asset, user=self.asset.owner
-        ).update(user=new_owner)
+        ).update(
+            user=new_owner,
+            counters=UpdateJSONFieldAttributes(
```
Comment (jamesrkiger) [inline kpi/deployment_backends/kobocat_backend.py:1294]: Ideally, I think it would be better to use a custom function that allows for removing these keys, rather than changing them to 0. But psql functions are still new territory for me, so it might take me a bit to figure out how. Let me know if you think it would be worth it.

File: kobo/apps/stripe/tests/test_one_time_addons_tracking.py [hunk 4]
```diff
@@ -0,0 +1,196 @@
...
+            'quantity': quantity,
+            **(product.metadata),
+        }
+        self.charge.save()
+
+    def test_increment_addon_usage(self):
+        # update_nlp_counter will error out if cache isn't updated first
+        # Also note that the used asr_seconds_current_month has to be > subscription limit,
+        # otherwise the addon will not be incremented (and neither will the counter, therefore)
+        self.organization.update_usage_cache(
```
Comment (jamesrkiger) [inline kobo/apps/stripe/tests/test_one_time_addons_tracking.py:152]: As I mention in the note in the code above, `update_nlp_counter()` does not work for asr/mt usage types if the org usage cache hasn't been updated previously. Unless I'm mistaken, the only time this cache gets updated currently is when the org usage endpoint is accessed. Should we initialize the org usage fields with data that can be used in the process of running `update_nlp_counter()` or should this function be changed to not depend only on this cache data?

File: kobo/apps/trackers/utils.py [hunk 5]
```diff
@@ -43,17 +46,36 @@ def update_nlp_counter(
 
     # Update the total counters by the usage amount to keep them current
     kwargs = {}
+    addon_amount_used = 0
     if service.endswith('asr_seconds'):
         kwargs['total_asr_seconds'] = F('total_asr_seconds') + amount
         if asset_id is not None:
             # If we're not updating the catch-all counter, increment any NLP add-ons the user may have
-            PlanAddOn.increment_add_ons_for_user(user_id, 'seconds', amount)
+            remaining = PlanAddOn.increment_add_ons_for_user(
```
Comment (jamesrkiger) [inline kobo/apps/trackers/utils.py:54]: `increment_add_ons_for_user()` gets a list of relevant addons and, for each addon, checks [is_organization_over_plan_limit()](https://github.com/kobotoolbox/kpi/blob/06b50e2712272d53f92cddfa9bcf4cb43b533613/kobo/apps/organizations/models.py#L77) for that organization. Maybe I'm missing a step somewhere, but it seems to me like this only applies to cases where a user has reached their plan limit at some point in the past. But consider a case where I have a plan limit of 1000 asr minutes and an addon for 2000 additional asr minutes, then I use 2000 asr minutes.  I would expect my addon minutes limits_remaining to go down to 1000. Currently it doesn't change at all. 

If you are curious about this, the [addon tracking tests](kobo/apps/stripe/tests/test_one_time_addons_tracking.py) above can be changed slightly to test it out.

Comment (jamesrkiger) [reply kobo/apps/trackers/utils.py:56]: Changed `is_organization_over_plan_limit` to `check_usage_exceeds_plan_limit`, which can check against incoming usage.

File: kobo/apps/stripe/tests/test_one_time_addons_tracking.py (hunk 4 above)
Comment (jamesrkiger) [reply kobo/apps/stripe/tests/test_one_time_addons_tracking.py:142]: Planning to address this issue in a subsequent PR.

**Review by bufke — CHANGES_REQUESTED** [review]
I asked about why we track addon_used_asr_seconds. @jamesrkiger is going to work on some documentation that will help me in this review. I'll leave these small comments for now and review again after.

**Comment by jamesrkiger** [comment]
I am closing this because we have decided not to implement tracking of addon usage over time (as opposed to simply incrementing addon usage when addon minutes/characters have been used). Some of this code may still be useful, but most of it is now irrelevant.
